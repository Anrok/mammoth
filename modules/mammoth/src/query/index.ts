export * from './base';
export * from './delete';
export * from './insert';
export * from './partial';
export * from './select';
export * from './update';
export * from './upsert';
