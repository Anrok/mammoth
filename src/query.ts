// This is used so we can capture the Returning of every different query type without having to
// specify every query seperately. The private property is used to simulate a nominal type so only
// this class is captured when doing a conditional type check (through T extends Query<infer Returning>).
export abstract class Query<Returning> {
  private _queryBrand!: Returning;
}