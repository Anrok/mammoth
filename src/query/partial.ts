import { GroupToken, ParameterToken, StringToken, Token, NoSpaceToken } from '../tokens';
import { Tokenable, Query } from './base';

// TODO: we can probably change this to be an expression.
export class PartialQuery implements Tokenable {
  tokens: Token[] = [];

  constructor(...tokens: Token[]) {
    this.tokens = tokens || [];
  }

  toTokens() {
    return this.tokens;
  }

  private add(partialQuery: PartialQuery | Query<any, any, any>, separator: string) {
    this.tokens.push(new StringToken(separator), ...partialQuery.toTokens());
    return this;
  }

  or(partialQuery: PartialQuery) {
    return this.add(partialQuery, `OR`);
  }

  and(partialQuery?: PartialQuery) {
    this.tokens.push(new StringToken(`AND`));

    if (partialQuery) {
      this.tokens.push(...partialQuery.toTokens());
    }
    return this;
  }

  andNotExists(partialQuery: PartialQuery | Query<any, any, any>) {
    this.tokens.push(new StringToken(`AND NOT EXISTS`), new GroupToken(partialQuery.toTokens()));
    return this;
  }

  andExists(partialQuery: PartialQuery | Query<any, any, any>) {
    this.tokens.push(new StringToken(`AND EXISTS`), new GroupToken(partialQuery.toTokens()));
    return this;
  }

  interval() {
    return this.castTo(`interval`);
  }

  castTo(type: string) {
    // TODO: should we enclose the already existing tokens in a group token so we surround it with
    // ( and ).
    this.tokens.push(new NoSpaceToken(`::${type}`));
    return this;
  }

  minus(string: string | PartialQuery) {
    if (string instanceof PartialQuery) {
      this.tokens.push(new StringToken(`-`), ...string.toTokens());
    } else {
      this.tokens.push(new StringToken(`-`), new ParameterToken(string));
    }

    return this;
  }

  plus(string: string) {
    this.tokens.push(new StringToken(`+`), new ParameterToken(string));
    return this;
  }

  gt(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`>`), ...partialQuery.toTokens());
    return this;
  }

  lt(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`<`), ...partialQuery.toTokens());
    return this;
  }

  gte(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`>=`), ...partialQuery.toTokens());
    return this;
  }

  lte(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`<=`), ...partialQuery.toTokens());
    return this;
  }

  not(partialQuery?: PartialQuery) {
    this.tokens.push(new StringToken(`NOT`));
    if (partialQuery) {
      this.tokens.push(...partialQuery.toTokens());
    }
    return this;
  }

  in(object: any[] | PartialQuery | Query<any, any, any>) {
    if (object instanceof PartialQuery) {
      this.tokens.push(new StringToken(`IN`), new GroupToken(object.tokens));
    } else if (object instanceof Query) {
      this.tokens.push(new StringToken(`IN`), new GroupToken(object.toTokens()));
    } else {
      this.tokens.push(new StringToken(`IN`), new GroupToken([new ParameterToken(object)]));
    }
    return this;
  }

  // TODO: Move this to an order by query.
  nullsFirst() {
    this.tokens.push(new StringToken(`NULLS FIRST`));
    return this;
  }

  nullsLast() {
    this.tokens.push(new StringToken(`NULLS LAST`));
    return this;
  }
}
