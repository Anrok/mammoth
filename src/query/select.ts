import { ColumnWrapper } from '../columns';
import { Database } from '../database';
import { Table } from '../table';
import { CollectionToken, SeparatorToken, StringToken } from '../tokens';
import { Tokenable, Query, QueryType } from './base';

export class SelectQuery<
  Db extends Database<any>,
  Row,
  InsertRow,
  UpdateRow,
  Ret,
  SingleRet,
  Tables = undefined
> extends Query<Db, Ret, SingleRet, Tables> {
  protected type: QueryType = 'ROWS';

  /** @internal */
  from<T extends Table<any, any, any>>(
    table: T,
  ): SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables & T['$row']> {
    // TODO: move this to the constructor so we do not have to expose SelectQuery#from().

    return this.internalFrom(table);
  }

  join<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'JOIN',
      table,
    );
  }

  crossJoin<T extends Table<any>>(table: T) {
    this.tokens.push(new StringToken(`CROSS JOIN`), new StringToken(table.getName()));
    return this;
  }

  innerJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'INNER JOIN',
      table,
    );
  }

  leftJoin<T extends Table<any>>(table: T) {
    // TODO: this isn't rock solid yet. It's changing the return type of the columns selected from the
    // left joined table to be undefined (as that's how it supposed to be). But, we're targeting the
    // columns based on the name, so a collision is likely and column#as() is not supported as it doesn't
    // change the return type.
    return this.internalJoin<
      T,
      SelectQuery<
        Db,
        Row,
        InsertRow,
        UpdateRow,
        // @ts-ignore
        (Pick<Ret[number], Exclude<keyof Ret[number], keyof T['$row']>> &
          Pick<
            { [P in keyof T['$row']]?: T['$row'][P] },
            // @ts-ignore
            Extract<keyof Ret[number], keyof T['$row']>
          >)[],
        Pick<SingleRet, Exclude<keyof SingleRet, keyof T['$row']>> &
          Pick<
            { [P in keyof T['$row']]?: T['$row'][P] },
            Extract<keyof SingleRet, keyof T['$row']>
          >,
        Tables
      >
    >('LEFT JOIN', table);
  }

  rightJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'RIGHT JOIN',
      table,
    );
  }

  leftOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'LEFT OUTER JOIN',
      table,
    );
  }

  rightOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'RIGHT OUTER JOIN',
      table,
    );
  }

  fullOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'FULL OUTER JOIN',
      table,
    );
  }

  fullJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'FULL JOIN',
      table,
    );
  }

  where(tokenable: Tokenable) {
    return this.internalWhere(tokenable);
  }

  limit(limit: number | 'ALL') {
    this.append`LIMIT`;

    if (typeof limit === 'number') {
      this.tokens.push(new StringToken(String(limit)));
    } else {
      this.tokens.push(new StringToken(`ALL`));
    }

    return this;
  }

  offset(offset: number) {
    this.tokens.push(new StringToken(`OFFSET ${offset}`));
    return this;
  }

  having(tokenable: Tokenable) {
    // TODO: should tokenable be a list like in orderBy?

    this.tokens.push(new StringToken(`HAVING`), ...tokenable.toTokens());
    return this;
  }

  orderBy(...tokenables: Tokenable[]) {
    this.tokens.push(
      new StringToken(`ORDER BY`),
      new SeparatorToken(
        `,`,
        tokenables.map(tokenable => new CollectionToken(tokenable.toTokens())),
      ),
    );
    return this;
  }

  forUpdate() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forNoKeyUpdate() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forShare() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forNoKeyShare() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  of(...tables: Table<any>[]) {
    this.tokens.push(
      new StringToken(`OF`),
      new SeparatorToken(
        `,`,
        tables.map(table => new StringToken(table.getName())),
      ),
    );
    return this;
  }

  skipLocked() {
    this.tokens.push(new StringToken(`SKIP LOCKED`));
    return this;
  }

  nowait() {
    this.tokens.push(new StringToken(`NOWAIT`));
    return this;
  }

  groupBy(...columns: ColumnWrapper<any, any, any, any, any>[]) {
    this.tokens.push(
      new StringToken(`GROUP BY`),
      new SeparatorToken(
        `,`,
        columns.map(column => new CollectionToken(column!.toTokens())),
      ),
    );
    return this;
  }
}
