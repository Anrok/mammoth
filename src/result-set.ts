import type { Column } from './column';
import { DeleteQuery } from './delete';
import type { Expression } from './expression';
import { InsertQuery } from './insert';
import { Query } from './query';
import { SelectQuery } from './select';
import { UpdateQuery } from './update';

export type ResultSetDataType<Type, IsNotNull extends boolean> = IsNotNull extends true ? Type : Type | null;

// This is not ideal, but, using dts-jest and it's snapshotting it's not capable to snapshot an e.g.
// optional number to `number | undefined`. Instead, it will snapshot to `number`. Because it's
// important to get the optional behaviour under test as well (it's so easy to create a regression)
// this flag is introduced to return a nominal class which gets snapshotted with the correct info.
export type ResultSet<T extends Query<any>> = T extends SelectQuery<
  infer Returning
>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends DeleteQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends UpdateQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends InsertQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : never;
