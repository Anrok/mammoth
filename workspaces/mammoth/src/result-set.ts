import { DbConfig, GetResultType } from './config';
import { Err, GetDataType } from './types';

import type { Column } from './column';
import { DeleteQuery } from './delete';
import type { Expression } from './expression';
import { InsertQuery } from './insert';
import { Query } from './query';
import { SelectQuery } from './select';
import { UpdateQuery } from './update';

export type ResultSetDataType<Config extends DbConfig, Type, IsNotNull> = IsNotNull extends true
  ? GetResultType<Config, Type>
  : GetResultType<Config, Type> | GetResultType<Config, 'Null'>;

type ReturningResultSet<Config extends DbConfig, Returning, Test extends boolean> = {
  [K in keyof Returning]: Returning[K] extends Column<
    any,
    any,
    any,
    infer D,
    infer N,
    any,
    infer JoinType
  >
    ? Extract<JoinType, 'left-join'> extends never
      ? Extract<JoinType, 'left-side-of-right-join'> extends never
        ? Extract<JoinType, 'full-join'> extends never
          ? N extends true
            ? Test extends true
              ? GetDataType<D, true>
              : ResultSetDataType<Config, D, true>
            : Test extends true
            ? GetDataType<D, false>
            : ResultSetDataType<Config, D, false>
          : Test extends true
          ? GetDataType<D, false>
          : ResultSetDataType<Config, D, false>
        : Test extends true
        ? GetDataType<D, false>
        : ResultSetDataType<Config, D, false>
      : Test extends true
      ? GetDataType<D, false>
      : ResultSetDataType<Config, D, false>
    : Returning[K] extends Expression<any, infer D, infer IsNotNull, any>
    ? Test extends true
      ? GetDataType<D, IsNotNull>
      : ResultSetDataType<Config, D, IsNotNull>
    : Returning[K] extends Query<{}>
    ? ResultSet<Config, Returning[K], Test>[keyof ResultSet<Config, Returning[K], Test>]
    : never;
};

// This is not ideal, but, using dts-jest and it's snapshotting it's not capable to snapshot an e.g.
// optional number to `number | undefined`. Instead, it will snapshot to `number`. Because it's
// important to get the optional behaviour under test as well (it's so easy to create a regression)
// this flag is introduced to return a nominal class which gets snapshotted with the correct info.
export type ResultSet<Config extends DbConfig, T, Test extends boolean> = T extends SelectQuery<
  Config,
  infer Returning
>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? Test extends true
                  ? GetDataType<D, true>
                  : ResultSetDataType<Config, D, true>
                : Test extends true
                ? GetDataType<D, false>
                : ResultSetDataType<Config, D, false>
              : Test extends true
              ? GetDataType<D, false>
              : ResultSetDataType<Config, D, false>
            : Test extends true
            ? GetDataType<D, false>
            : ResultSetDataType<Config, D, false>
          : Test extends true
          ? GetDataType<D, false>
          : ResultSetDataType<Config, D, false>
        : // To avoid an issue where I (IsNotNull) would become boolean instead of true or false, we
        // extend Expression<..> twice with separate infers.
        Returning[K] extends Expression<any, any, infer I, string>
        ? Returning[K] extends Expression<any, infer D, I, string>
          ? Test extends true
            ? GetDataType<D, I>
            : ResultSetDataType<Config, D, I>
          : Err<'not an expression'>
        : Returning[K] extends Query<{}>
        ? ResultSet<Config, Returning[K], Test>[keyof ResultSet<Config, Returning[K], Test>]
        : never;
    }
  : T extends DeleteQuery<Config, any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        Config,
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? Test extends true
                  ? GetDataType<D, true>
                  : ResultSetDataType<Config, D, true>
                : Test extends true
                ? GetDataType<D, false>
                : ResultSetDataType<Config, D, false>
              : Test extends true
              ? GetDataType<D, false>
              : ResultSetDataType<Config, D, false>
            : Test extends true
            ? GetDataType<D, false>
            : ResultSetDataType<Config, D, false>
          : Test extends true
          ? GetDataType<D, false>
          : ResultSetDataType<Config, D, false>
        : Returning[K] extends Expression<any, infer D, infer IsNotNull, any>
        ? Test extends true
          ? GetDataType<D, IsNotNull>
          : ResultSetDataType<Config, D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Config, Returning[K], Test>[keyof ResultSet<Config, Returning[K], Test>]
        : never;
    }
  : T extends UpdateQuery<Config, any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        Config,
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? Test extends true
                  ? GetDataType<D, true>
                  : ResultSetDataType<Config, D, true>
                : Test extends true
                ? GetDataType<D, false>
                : ResultSetDataType<Config, D, false>
              : Test extends true
              ? GetDataType<D, false>
              : ResultSetDataType<Config, D, false>
            : Test extends true
            ? GetDataType<D, false>
            : ResultSetDataType<Config, D, false>
          : Test extends true
          ? GetDataType<D, false>
          : ResultSetDataType<Config, D, false>
        : Returning[K] extends Expression<any, infer D, infer IsNotNull, any>
        ? Test extends true
          ? GetDataType<D, IsNotNull>
          : ResultSetDataType<Config, D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Config, Returning[K], Test>[keyof ResultSet<Config, Returning[K], Test>]
        : never;
    }
  : T extends InsertQuery<Config, any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? Test extends true
                  ? GetDataType<D, true>
                  : ResultSetDataType<Config, D, true>
                : Test extends true
                ? GetDataType<D, false>
                : ResultSetDataType<Config, D, false>
              : Test extends true
              ? GetDataType<D, false>
              : ResultSetDataType<Config, D, false>
            : Test extends true
            ? GetDataType<D, false>
            : ResultSetDataType<Config, D, false>
          : Test extends true
          ? GetDataType<D, false>
          : ResultSetDataType<Config, D, false>
        : Returning[K] extends Expression<any, infer D, infer IsNotNull, any>
        ? Test extends true
          ? GetDataType<D, IsNotNull>
          : ResultSetDataType<Config, D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Config, Returning[K], Test>[keyof ResultSet<Config, Returning[K], Test>]
        : never;
    }
  : Err<'not a query'>;
