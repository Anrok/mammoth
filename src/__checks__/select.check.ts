import {
  arrayAgg,
  coalesce,
  count,
  defineDb,
  defineTable,
  integer,
  sum,
  text,
  timestampWithTimeZone,
  uuid,
} from '../../.build';

import { Query } from '../../.build/query';
import { ResultSet } from '../../.build/result-set';

const toSnap = <T extends Query<any>>(query: T): ResultSet<T, true> => {
  return undefined as any;
};

/** @dts-jest enable:test-type */

const foo = defineTable({
  id: uuid().primaryKey().default(`gen_random_uuid()`),
  createDate: timestampWithTimeZone().notNull().default(`now()`),
  name: text().notNull(),
  value: integer(),
});

const bar = defineTable({
  id: uuid().primaryKey().default(`gen_random_uuid()`),
  startDate: timestampWithTimeZone().notNull().default(`now()`),
  endDate: timestampWithTimeZone().notNull().default(`now()`),
  value: integer(),
  fooId: uuid().references(foo, 'id'),
});

const db = defineDb({ foo, bar }, () => Promise.resolve({ rows: [], affectedCount: 0 }));

// @dts-jest:group select
{
  // @dts-jest:snap should return null and not null properties
  toSnap(db.select(db.foo.id, db.foo.createDate, db.foo.value).from(db.foo));

  // @dts-jest:snap should return nullable properties of left joined columns
  toSnap(db.select(db.foo.id, db.bar.endDate, db.bar.value).from(db.foo).leftJoin(db.bar));

  // @dts-jest:snap should return nullable properties of left side properties when right joining
  toSnap(db.select(db.foo.name, db.bar.startDate, db.bar.value).from(db.foo).rightJoin(db.bar));

  // @dts-jest:snap should return renamed properties because of alias
  toSnap(db.select(db.foo.name.as(`fooName`), db.foo.value.as(`fooValue`)).from(db.foo));

  // @dts-jest:snap should return nullable properties of all sides because of full join
  toSnap(db.select(db.foo.name, db.bar.startDate, db.bar.value).from(db.foo).fullJoin(db.bar));

  // @dts-jest:snap should select expression
  toSnap(db.select(db.foo.value.plus(1)).from(db.foo));

  // @dts-jest:snap should select named expression
  toSnap(db.select(db.foo.value.plus(1).as(`test`)).from(db.foo));

  // @dts-jest:snap should select aggregate subquery
  toSnap(db.select(db.foo.id, db.select(count()).from(db.foo)).from(db.foo));

  // @dts-jest:snap should select array_agg
  toSnap(db.select(arrayAgg(db.foo.name)).from(db.foo));

  // @dts-jest:snap should select null column in subquery
  toSnap(db.select(db.foo.id, db.select(db.foo.value).from(db.foo)).from(db.foo));

  // @dts-jest:snap should select aggregate with alias
  toSnap(db.select(db.foo.id, sum(db.foo.value).as(`total`)).from(db.foo));

  // @dts-jest:snap should convert null value to not null using coalesce
  toSnap(db.select(coalesce(db.foo.value, 1)).from(db.foo));

  toSnap(
    db
      .select(db.foo.id)
      .from(db.foo)
      // @dts-jest:fail:snap should not use in with wrong data type
      .where(db.foo.id.in(db.select(db.foo.createDate).from(db.foo))),
  );

  // @dts-jest:snap should select case with correct type and alias
  toSnap(
    db
      .select(
        db
          .case()
          .when(db.foo.value.gt(100))
          .then('A' as const)
          .when(db.foo.value.gt(0))
          .then('B' as const)
          .else('C' as const)
          .end()
          .as(`bar`),
      )
      .from(db.foo),
  );

  db.select(db.foo.id, db.foo.value)
    .from(db.foo)
    .then((result) => {
      // @dts-jest:snap should select and await result set
      result;
    });
}
