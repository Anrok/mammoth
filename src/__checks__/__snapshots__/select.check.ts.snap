// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select should convert null value to not null using coalesce (type) should match snapshot 1`] = `"{ coalesce: GetDataType<number, true>; }"`;

exports[`select should not use in with wrong data type (type) should match snapshot 1`] = `
"Argument of type 'SelectQuery<GetSelectable<Column<\\"createDate\\", \\"foo\\", Date, true, any, undefined>>>' is not assignable to parameter of type 'string[] | Expression<string, true, any>'.
  Type 'SelectQuery<GetSelectable<Column<\\"createDate\\", \\"foo\\", Date, true, any, undefined>>>' is missing the following properties from type 'Expression<string, true, any>': name, nameIsAlias, _expressionBrand, unusedName, and 33 more."
`;

exports[`select should return null and not null properties (type) should match snapshot 1`] = `"{ id: GetDataType<string, true>; createDate: GetDataType<Date, true>; value: GetDataType<number, false>; }"`;

exports[`select should return nullable properties of all sides because of full join (type) should match snapshot 1`] = `"{ name: GetDataType<string, false>; startDate: GetDataType<Date, false>; value: GetDataType<number, false>; }"`;

exports[`select should return nullable properties of left joined columns (type) should match snapshot 1`] = `"{ id: GetDataType<string, true>; endDate: GetDataType<Date, false>; value: GetDataType<number, false>; }"`;

exports[`select should return nullable properties of left side properties when right joining (type) should match snapshot 1`] = `"{ name: GetDataType<string, false>; startDate: GetDataType<Date, true>; value: GetDataType<number, false>; }"`;

exports[`select should return renamed properties because of alias (type) should match snapshot 1`] = `"{ fooName: GetDataType<string, true>; fooValue: GetDataType<number, false>; }"`;

exports[`select should select aggregate subquery (type) should match snapshot 1`] = `"{ id: GetDataType<string, true>; count: GetDataType<string, true>; }"`;

exports[`select should select aggregate with alias (type) should match snapshot 1`] = `"{ id: GetDataType<string, true>; total: GetDataType<number, false>; }"`;

exports[`select should select and await result set (type) should match snapshot 1`] = `"{ id: string; value: number; }[]"`;

exports[`select should select array_agg (type) should match snapshot 1`] = `"{ arrayAgg: GetDataType<string[], false>; }"`;

exports[`select should select case with correct type and alias (type) should match snapshot 1`] = `"{ bar: GetDataType<\\"A\\" | \\"B\\" | \\"C\\", true>; }"`;

exports[`select should select expression (type) should match snapshot 1`] = `"{ \\"?column?\\": GetDataType<number, false>; }"`;

exports[`select should select named expression (type) should match snapshot 1`] = `"{ test: GetDataType<number, false>; }"`;

exports[`select should select null column in subquery (type) should match snapshot 1`] = `"{ id: GetDataType<string, true>; value: GetDataType<number, false>; }"`;
